### PR TITLE
alternative (easier?) way to define datasets

### PR DESCRIPTION
**Pull request by richardjgowers**
@orbeckst what do you think about this?  I was looking at how to write tests for this package, and with every single dataset implementing how it should be downloaded, there was a lot of room for potential errors.  Instead if we define a template for how a dataset should look, we can have the logic of how this is retrieved in a centralised location.

It takes ideas from how we auto-register Readers in MDAnalysis, so Datasets are added to a list of available datasets once defined.

### CODE REVIEW

File: MDAnalysisData/adk_equilibrium.py [hunk 1]
```diff
@@ -5,88 +5,32 @@
...
-
-    module_path = dirname(__file__)
-    with codecs.open(join(module_path, 'descr', DESCRIPTION),
-                     encoding="utf-8") as dfile:
-        records.DESCR = dfile.read()
-
-    return records
+class ADK_Equilibrium(Dataset):
+    NAME = "adk_equilibrium"
+    DESCRIPTION = "adk_equilibrium.rst"
```
Comment (richardjgowers) [inline MDAnalysisData/adk_equilibrium.py]: One downside I can see is we've lost the short description that the function to generate this had

File: MDAnalysisData/base.py [hunk 2]
```diff
@@ -94,6 +96,59 @@ def __setstate__(self, state):
...
+        with codecs.open(join(module_path, 'descr', self.DESCRIPTION),
+                         encoding="utf-8") as dfile:
+            contents['DESCR'] = dfile.read()
+
+
+        # finally, init the Bunch object
+        super().__init__(**contents)
+
+
+def fetch(dataset, data_home=None, download_if_missing=True):
```
Comment (richardjgowers) [inline MDAnalysisData/base.py]: This allows `MDADATA.fetch('adk_equilibrium')`, rather than a separate function for each dataset

File: MDAnalysisData/adk_equilibrium.py [hunk 3]
```diff
@@ -5,88 +5,32 @@
...
-        Specify another download and cache folder for the datasets. By default
-        all MDAnalysisData data is stored in '~/MDAnalysis_data' subfolders.
-        This dataset is stored in ``<data_home>/adk_equilibrium``.
-    download_if_missing : optional, default=True
-        If ``False``, raise a :exc:`IOError` if the data is not locally available
-        instead of trying to download the data from the source site.
-
-    Returns
-    -------
-    dataset : dict-like object with the following attributes:
```
Comment (richardjgowers) [inline MDAnalysisData/adk_equilibrium.py]: and another thing lost is the clear description of what the returned Bunch will have....

File: MDAnalysisData/base.py (hunk 2 above)
Comment (orbeckst) [reply MDAnalysisData/base.py]: There's a reason why explicit functions: tab completion and introspection. (sklearn does it and it works really well – much better than having to know the name of the dataset)

I'd like to keep explicit functions – both for ease of use and for same "look and feel" as sklearn.datasets (as well as getting docs!)

We can have a generic mechanism and generate the `fetch_*` functions.

File: MDAnalysisData/adk_equilibrium.py (hunk 3 above)
Comment (orbeckst) [reply MDAnalysisData/adk_equilibrium.py]: I don't want to loose this...

File: MDAnalysisData/adk_equilibrium.py (hunk 1 above)
Comment (orbeckst) [reply MDAnalysisData/adk_equilibrium.py]: Don't want to loose the description and don't want to loose the docs...

File: MDAnalysisData/adk_equilibrium.py [hunk 4]
```diff
@@ -5,88 +5,43 @@
...
-    return records
+    NAME = "adk_equilibrium"
+    DESCRIPTION = "adk_equilibrium.rst"
+
+    # The original data can be found at the figshare URL.
+    # The SHA256 checksum of the zip file changes with every download so we
+    # cannot check its checksum. Instead we download individual files.
+    # separately. The keys of this dict are also going to be the keys in the
+    # Bunch that is returned.
+    ARCHIVE = {
```
Comment (orbeckst) [inline MDAnalysisData/adk_equilibrium.py]: If we normalize all of this then we might be able to just put all these data into JSON or YAML files.

File: MDAnalysisData/base.py (hunk 2 above)
Comment (orbeckst) [reply MDAnalysisData/base.py]: Yes, that's nice for this case. But did you look at some of the other accessors like `fetch_adk_transitions_DIMS` where we get a tar file and unpack? We might be able to reduce our requirements to these two types of cases.

**Comment by orbeckst** [comment]
It's a good idea to have a template for the most common types of accessing
- separate files
- archives that need to be unpacked

I still want to keep explicit accessor functions because of ease of use, docs, and keeping similarity to sklearn.datasets.

We still need to be able to allow other code that does not fit into the general templates.

File: MDAnalysisData/base.py (hunk 2 above)
Comment (orbeckst) [reply MDAnalysisData/base.py]: `RemoteFileMetadata` is verbatim from sklearn. Might be useful to keep it that way and really keep it simple. 

If anything, we should build data structures that contain `RemoteFileMetadata` instances that map remote and local. Have a look at the transitions dataset to see what else we have. 

Finally, have a look at [sklearn.datasets](https://github.com/scikit-learn/scikit-learn/tree/master/sklearn/datasets) (and the outstanding [docs](http://scikit-learn.org/stable/datasets/index.html)) to see the variance. I think one reason for copy&paste code is that ultimately each dataset in the wild might have slightly different requirements. Still, that's not to say that we can't try to get a bit of order in ;-).

**Comment by orbeckst** [comment]
Might be worthwhile reviving this PR.... if someone wants to look into it.
